go_lib/hooks/tls_certificate/internal_tls.go

```diff
@@ -0,0 +1,267 @@
+// Register hook which save tls cert in values from secret.
+// If secret is not created hook generate CA with long expired time
+// and generate tls cert for passed domains signed with generated CA.
+// That CA cert and TLS cert and private key MUST save in secret with helm.
+// Otherwise, every d8 restart will generate new tls cert.
+// Tls cert also has long expired time same as CA 87600h == 10 years.
+// Therese tls cert often use for in cluster https communication
+// with service which order tls
+// Clients need to use CA cert for verify connection
+func RegisterInternalTLSHook(conf GenSelfSignedTLSHookConf) bool {
+	return sdk.RegisterFunc(&go_hook.HookConfig{
+		OnBeforeHelm: &go_hook.OrderedConfig{Order: 5},
+		Kubernetes: []go_hook.KubernetesConfig{
+			{
+				Name:       "secret",
+				ApiVersion: "v1",
+				Kind:       "Secret",
+				NamespaceSelector: &types.NamespaceSelector{
+					NameSelector: &types.NameSelector{
+						MatchNames: []string{conf.Namespace},
+					},
+				},
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{conf.TLSSecretName},
+				},
+
+				FilterFunc: tlsFilter,
+			},
+		},
+	}, genSelfSignedTLS(conf))
+}
+
+func tlsFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var secret v1.Secret
+	err := sdk.FromUnstructured(obj, &secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return certificate.Certificate{
+		CA:   string(secret.Data["ca.crt"]),
+		Cert: string(secret.Data["tls.crt"]),
+		Key:  string(secret.Data["tls.key"]),
+	}, nil
+}
+
+func genSelfSignedTLS(conf GenSelfSignedTLSHookConf) func(input *go_hook.HookInput) error {
+	caPath, certPath, keyPath := conf.generatePaths()
+
+	return func(input *go_hook.HookInput) error {
+		var cert certificate.Certificate
+		var err error
+
+		cn, sans := conf.CN, conf.SANs(input)
+
+		if len(input.Snapshots["secret"]) == 0 {
+			// No certificate in snapshot => generate a new one.
+			// Secret will be updated by Helm.
+			cert, err = generateNewTLS(input, cn, sans)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Certificate is in the snapshot => load it.
+			cert = input.Snapshots["secret"][0].(certificate.Certificate)
+			// update certificate if less than 6 month left. We create certificate for 10 years, so it looks acceptable
+			// and we don't need to create Crontab schedule
+			caOutdated, err := isOutdatedCA(cert.CA)
+			if err != nil {
+				return err
+			}
+			certOutdated, err := isOutdatedCert(cert.Cert, sans)
+			if err != nil {
+				return err
+			}
+
+			if caOutdated || certOutdated {
+				cert, err = generateNewTLS(input, cn, sans)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// Note that []byte values will be encoded in base64. Use strings here!
+		input.Values.Set(caPath, cert.CA)
+		input.Values.Set(certPath, cert.Cert)
+		input.Values.Set(keyPath, cert.Key)
+		return nil
+	}
+}
+
+func isOutdatedCert(certData string, desiredSANSs []string) (bool, error) {
```

Comment:
From the function's name, devs can guess that it only checks for the expiration of a certificate, which is not valid. Let's consider changing the name to something more suitable.